scw/client.go

```diff
@@ -340,6 +348,103 @@ func (c *Client) doListAll(req *ScalewayRequest, res interface{}) (err error) {
 }
 
+// doListZones collects all zones using multiple list requests and aggregate all results on a single response.
+// result is sorted by zone
+func (c *Client) doListZones(req *ScalewayRequest, res interface{}, zones []Zone) (err error) {
+	if response, isLister := res.(lister); isLister {
+		// Prepare request with %zone% that can be replaced with actual zone
+		path := req.Path
+		for _, zone := range AllZones {
+			if strings.Contains(req.Path, string(zone)) {
+				path = strings.ReplaceAll(req.Path, string(zone), "%zone%")
+				break
+			}
+		}
+		if !strings.Contains(path, "%zone%") {
+			return fmt.Errorf("request is not a valid zoned request")
+		}
+
+		// Requests are parallelized
+		responseMutex := sync.Mutex{}
+		requestGroup := sync.WaitGroup{}
+		errChan := make(chan error, len(zones))
+
+		requestGroup.Add(len(zones))
+		for _, zone := range zones {
+			go func(zone Zone) {
+				defer requestGroup.Done()
+				// Request is cloned as doListAll will change header
+				// We remove zones as it would recurse in the same function
+				req := req.clone()
+				req.zones = []Zone(nil)
+				req.Path = strings.ReplaceAll(path, "%zone%", string(zone))
+
+				// We create a new response that we append to main response
+				zoneResponse := newVariableFromType(response)
+				err := c.Do(req, zoneResponse)
+				if err != nil {
+					errChan <- err
+				}
+				responseMutex.Lock()
+				_, err = response.UnsafeAppend(zoneResponse)
+				responseMutex.Unlock()
+				if err != nil {
+					errChan <- err
+				}
+			}(zone)
+		}
+		requestGroup.Wait()
+
+	L: // We gather potential errors and return them all together
+		for {
+			select {
+			case newErr := <-errChan:
+				err = errors.Wrap(err, newErr.Error())
+			default:
+				break L
+			}
+		}
+		close(errChan)
+		if err != nil {
+			return err
+		}
+
+		sortResponseByZone(res)
+		return nil
+	}
+
+	return errors.New("%T does not support pagination", res)
+}
+
+// sortSliceByZone sorts a slice of struct using a Zone field that should exist
+func sortSliceByZone(list interface{}) {
+	listValue := reflect.ValueOf(list)
+	sort.Slice(list, func(i, j int) bool {
+		zone1 := listValue.Index(i).Elem().FieldByName("Zone").Interface().(Zone)
+		zone2 := listValue.Index(j).Elem().FieldByName("Zone").Interface().(Zone)
+		return zone1 < zone2
+	})
+}
+
+// sortResponseByZone find first field that is a slice in a struct and sort it by zone
+func sortResponseByZone(res interface{}) {
+	// res may be ListServersResponse
+	//
+	// type ListServersResponse struct {
+	//	TotalCount uint32 `json:"total_count"`
+	//	Servers []*Server `json:"servers"`
+	// }
+	// We iterate over fields searching for the slice one to sort it
+	resType := reflect.TypeOf(res).Elem()
+	fields := reflect.VisibleFields(resType)
+	for _, field := range fields {
+		if field.Type.Kind() == reflect.Slice {
+			sortSliceByZone(reflect.ValueOf(res).Elem().FieldByName(field.Name).Interface())
```

Comment:
We should respect the sort asked by the user

Reply:
Fixed by using stable sort, user's sort will be respected but result will still be sorted by locality